catroid/src/androidTest/java/org.catrobat.catroid/uiespresso/util/matchers/AdapterViewMatchers.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Catroid: An on-device visual programming system for Android devices
+ * Copyright (C) 2010-2017 The Catrobat Team
+ * (<http://developer.catrobat.org/credits>)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * An additional term exception under section 7 of the GNU Affero
+ * General Public License, version 3, is available at
+ * http://developer.catrobat.org/license_additional_term
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.catrobat.catroid.uiespresso.util.matchers;
+
+import android.view.View;
+import android.widget.Adapter;
+import android.widget.AdapterView;
+
+import org.catrobat.catroid.common.ProjectData;
+import org.catrobat.catroid.ui.adapter.ProjectListAdapter;
+import org.hamcrest.Description;
+import org.hamcrest.Matcher;
+import org.hamcrest.TypeSafeMatcher;
+
+public final class AdapterViewMatchers {
+	private AdapterViewMatchers() {
+		throw new AssertionError();
+	}
+
+	public static Matcher<View> withAdaptedDataAt(final String expectedContent, final int listPosition) {
+		return new TypeSafeMatcher<View>() {
+
+			@Override
+			public void describeTo(Description description) {
+			}
+
+			@Override
+			public boolean matchesSafely(View view) {
+				if (!(view instanceof AdapterView)) {
+					return false;
+				}
+
+				Adapter adapter = ((AdapterView) view).getAdapter();
+				String listItem;
+				if (adapter instanceof ProjectListAdapter) {
+					listItem = ((ProjectData) adapter.getItem(listPosition)).projectName;
+				} else {
+					listItem = (String) adapter.getItem(listPosition);
+				}
+
+				return expectedContent.equals(listItem);
+			}
+		};
+	}
+
+	public static Matcher<View> withIndex(final Matcher<View> matcher, final int index) {
+		return new TypeSafeMatcher<View>() {
+			int currentIndex = 0;
+
+			@Override
+			public void describeTo(Description description) {
+				description.appendText("with index: ");
+				description.appendValue(index);
+				matcher.describeTo(description);
+			}
+
+			@Override
+			public boolean matchesSafely(View view) {
+				return matcher.matches(view) && currentIndex++ == index;
```

Comment:
but you do have an adapter view, and should use onData in the first place, instead of writing a matcher that does something that espresso can do for you out of the box.
the stackoverflow answer has some stuff not in a adapterView and uses this approach because of this.

write a matcher that ensures you got the right list on your hands similar to this one here:
https://github.com/Catrobat/Catroid/blob/develop/catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/util/matchers/BrickCategoryListMatchers.java

and then use it like this:

onData(allOf(is(instanceOf(String.class)), is(UiTestUtils.getResourcesString(categoryStringResource))))
					.inAdapterView(BrickCategoryListMatchers.isBrickCategoryView())
					.check(matches(isDisplayed()));

found in https://github.com/Catrobat/Catroid/blob/develop/catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/BrickValueParameterTest.java

or even add a wrapper that calls above chunk of code and gives you a data interaction like this:
public static BrickDataInteractionWrapper onBrickAtPosition(int position) {
		return new BrickDataInteractionWrapper(onData(instanceOf(Brick.class))
				.inAdapterView(ScriptListMatchers.isScriptListView())
				.atPosition(position));
	}

https://github.com/Catrobat/Catroid/blob/develop/catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/utils/BrickDataInteractionWrapper.java

and when using espresso functionality instead of your own matcher that does "atPosition", "inAdapterView", "onChildView", etc...


and then